test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineRetransform/libRedefineRetransform.cpp

```diff
@@ -0,0 +1,294 @@
+
+            *newClassBytes = buf;
+            *newLen = len;
+
+            _log("  ClassFileLoadHook: set new class bytes\n");
+        }
+        return true;
+    }
+
+    void setSavedHookClassBytes(const unsigned char* bytes, jint len) {
+        jvmtiError err = jvmti->Allocate(len, &savedClassBytes);
+        if (err != JVMTI_ERROR_NONE) {
+            _log("ClassFileLoadHook: failed to allocate %ld bytes for saved class bytes: %d", len, err);
+            return;
+        }
+        memcpy(savedClassBytes, bytes, len);
+        savedClassBytesLen = len;
+    }
+
+    jbyteArray getSavedHookClassBytes() {
+        if (savedClassBytes == nullptr) {
+            _log("%s: savedClassBytes is NULL\n", mode);
+            return nullptr;
+        }
+
+        jbyteArray result = env->NewByteArray(savedClassBytesLen);
+        if (result == nullptr) {
+            _log("%s: NewByteArray(%ld) failed\n", mode, savedClassBytesLen);
+        } else {
+            jbyte* arrayPtr = env->GetByteArrayElements(result, nullptr);
+            if (arrayPtr == nullptr) {
+                _log("%s: Failed to get array elements\n", mode);
+                result = nullptr;
+            } else {
+                memcpy(arrayPtr, savedClassBytes, savedClassBytesLen);
+                env->ReleaseByteArrayElements(result, arrayPtr, JNI_COMMIT);
+            }
+        }
+        return result;
+    }
+};
+
+ClassFileLoadHookHelper* ClassFileLoadHookHelper::instance = nullptr;
+
+
+JNIEXPORT void JNICALL
+callbackClassFileLoadHook(jvmtiEnv *jvmti_env,
+        JNIEnv* jni_env,
+        jclass class_being_redefined,
+        jobject loader,
+        const char* name,
+        jobject protection_domain,
+        jint class_data_len,
+        const unsigned char* class_data,
+        jint* new_class_data_len,
+        unsigned char** new_class_data) {
+    if (isTestClass(name)) {
+        _log(">>ClassFileLoadHook: %s, %ld bytes, ptr = %p\n", name, class_data_len, class_data);
+
+        ClassFileLoadHookHelper* helper = ClassFileLoadHookHelper::getInstance();
+        if (helper == nullptr) {
+            _log("ClassFileLoadHook ERROR: helper instance is not initialized\n");
+            return;
+        }
+        // save class bytes
+        helper->setSavedHookClassBytes(class_data, class_data_len);
+        // set new class bytes
+        helper->getHookClassBytes(new_class_data, new_class_data_len);
+
+        _log("<<ClassFileLoadHook\n");
+    }
+}
+
+JNIEXPORT jint JNICALL Agent_OnLoad(JavaVM* jvm, char* options, void* reserved) {
+    jint res = jvm->GetEnv((void **)&jvmti, JVMTI_VERSION_1_1);
+    if (res != JNI_OK) {
+        _log("Failed to get JVMTI interface: %ld\n", res);
+        return JNI_ERR;
+    }
+
+    jvmtiCapabilities caps;
+    memset(&caps, 0, sizeof(caps));
+
+    caps.can_redefine_classes = 1;
+    caps.can_retransform_classes = 1;
+    res = jvmti->AddCapabilities(&caps);
+    if (res != JVMTI_ERROR_NONE) {
```

Comment:
The type of res is jint. It is better to use a variable of type jvmtiError. I've missed this initially.

Reply:
Fixed